hoomd/Autotuner.h

```diff
@@ -24,227 +34,620 @@
+        n_parameters_max *= dimension.size();
+        }
+
+    // Populate the array of samples.
+    m_parameters.reserve(n_parameters_max);
+    std::array<unsigned int, n_dimensions> placeholder;
+    placeholder.fill(0);
+    initializeParameters(dimension_ranges, placeholder, is_parameter_valid, 0);
+
+    if (m_parameters.size() == 0)
+        {
+        std::ostringstream s;
+        s << "Autotuner " << m_name << " has no valid parameters.";
+        throw std::invalid_argument(s.str());
+        }
+
+    // Ensure that m_n_samples is non-zero and odd to simplify the median computation.
+    if ((m_n_samples & 1) == 0)
+        m_n_samples += 1;
+
+    // Initialize memory.
+    m_samples.resize(m_parameters.size());
+    m_sample_median.resize(m_parameters.size());
+
+    for (unsigned int i = 0; i < m_parameters.size(); i++)
+        {
+        m_samples[i].resize(m_n_samples);
+        }
+
+// create CUDA events
+#ifdef ENABLE_HIP
+    hipEventCreate(&m_start);
+    hipEventCreate(&m_stop);
+    CHECK_CUDA_ERROR();
+#endif
+
+    m_current_element = 0;
+    m_current_sample = 0;
+    m_current_param = m_parameters[m_current_element];
+    if (optional)
+        {
+        m_state = INACTIVE;
+        }
+    else
+        {
+        m_state = SCANNING;
+        }
+    }
+
+template<size_t n_dimensions> void Autotuner<n_dimensions>::end()
+    {
+#ifdef ENABLE_HIP
+    // handle timing updates if scanning
+    if (m_state == SCANNING)
+        {
+        hipEventRecord(m_stop, 0);
+        hipEventSynchronize(m_stop);
+        hipEventElapsedTime(&m_samples[m_current_element][m_current_sample], m_start, m_stop);
+
+        m_exec_conf->msg->notice(9)
+            << "Autotuner " << m_name << ": t[" << formatParam(m_current_param) << ","
+            << m_current_sample << "] = " << m_samples[m_current_element][m_current_sample]
+            << std::endl;
+
+        if (this->m_exec_conf->isCUDAErrorCheckingEnabled())
+            CHECK_CUDA_ERROR();
+        }
+#endif
+
+    // Handle state data updates and transitions.
+    if (m_state == SCANNING)
+        {
+        // move on to the next element
+        m_current_element++;
+
+        // If we hit the end of the elements
+        if (m_current_element >= m_parameters.size())
+            {
+            // Move on to the next sample.
+            m_current_sample++;
+            m_current_element = 0;
+
+            // If this is the last sample, go to the idle state and compute the optimal parameter.
+            if (m_current_sample >= m_n_samples)
+                {
+                m_state = IDLE;
+                m_current_sample = 0;
+                m_current_param = m_parameters[computeOptimalParameterIndex()];
+                }
+            else
+                {
+                m_current_param = m_parameters[m_current_element];
+                }
+            }
+        else
+            {
+            m_current_param = m_parameters[m_current_element];
+            }
```

Comment:
No. It is an error, in fact. It will overwrite a `m_current_param` set by the user. It is probably left over from the old `Autotuner` code which had a more complex state machine.

Reply:
On looking at this more carefully, the code is correct as implemented. It does not override the parameter set by the user because this is in an `if (m_state == SCANNING)` block.

Reply:
The reason this is necessary in the first place is because of the order in when the state machine moves over elements and samples.